src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -272,14 +271,129 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
+       *                    \___________________  | |  __________________________/    |
+       *                                        \ | | /                               |
+       *                     ---------------->    Node                              Table
+       *                    /                      |
+       *                   /                      ...
+       *       ___________/_________
+       *   The diamond's origin node has four outputs, but only three outputs are part of the diamond structure.
+       *   Therefore, we do not want to continue the pushdown traversal below the diamond. Because otherwise, we would
+       *   incorrectly filter the Join's left input.
+       *
+       * To identify cases such as above, we check the outputs count of the diamond's origin node and compare it
+       * with the number of UnionNodes in the diamond structure.
+       */
+      size_t union_node_count = 0;
+      visit_lqp(union_node, [&](const auto& diamond_node) {
+        if (diamond_node == diamond_origin_node) return LQPVisitation::DoNotVisitInputs;
+        if (diamond_node->type == LQPNodeType::Union) union_node_count++;
+        return LQPVisitation::VisitInputs;
+      });
+      if (diamond_origin_node->output_count() != union_node_count + 1) {
+        handle_barrier();
+        return;
+      }
+
+      if (diamond_origin_node->input_count() == 1) {
+        auto diamond_push_down_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>{};
+        const auto node_below_diamond = diamond_origin_node->left_input();
+        for (const auto& push_down_node : push_down_nodes) {
+          if (node_below_diamond && _is_evaluable_on_lqp(push_down_node, node_below_diamond)) {
+            // Push predicate below the diamond's origin node
+            diamond_push_down_nodes.emplace_back(push_down_node);
+          } else if (_is_evaluable_on_lqp(push_down_node, diamond_origin_node)) {
+            // Push predicate above the diamond's origin node
+            lqp_insert_node_above(diamond_origin_node, push_down_node, AllowRightInput::Yes);
+          } else {
+            // Insert predicate above the diamond, so that it stays evaluable.
+            lqp_insert_node_above(union_node, push_down_node, AllowRightInput::Yes);
+          }
+        }
+        // Continue predicate pushdown below the diamond structure
+        _push_down_traversal(diamond_origin_node, LQPInputSide::Left, diamond_push_down_nodes, estimator);
+
+      } else {
+        // Do not move pushdown predicates below the diamond's origin node if it is a Join, a Union or another node
+        // with multiple inputs. Instead, insert the pushdown predicates above it and call the _push_down_traversal
+        // subroutine again, so that the predicate pushdown for the node with multiple inputs is handled appropriately.
+        auto updated_diamond_origin_node = diamond_origin_node;
```

Comment:
`auto&`?

Reply:
Not possible here since we want to update the variable later, if necessary.

Reply:
Not talking about constness here, only about using a reference and not taking the ownership of the shared pointer(s). But I guess we're not consistent regarding this in the code base at all.

Reply:
Ok, but I think this if will fail then:

```cpp
if (updated_diamond_origin_node == diamond_origin_node) {
```

Reply:
Indeed. The compiler requires `diamond_origin_node` to be non-const. As a result, we would change `diamond_origin_node`, if we assign sth. to `updated_diamond_origin_node` - compromising the above if stmt.

Reply:
Got it, did not think about this